public/js/lib/devtools-sham/transport/ws-transport.js

```diff
@@ -0,0 +1,51 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+const EventEmitter = require("devtools-sham/shared/event-emitter");
+
+function WebSocketDebuggerTransport(socket) {
+  EventEmitter.decorate(this);
+
+  this._ws = socket;
+
+  this.active = false;
+  this.hooks = null;
+}
+
+WebSocketDebuggerTransport.prototype = {
+  ready() {
+    if (!this.active) {
+      this.active = true;
+      this._ws.onmessage = this._onMessage.bind(this);
+    }
+  },
+
+  send(object) {
+    this.emit("send", object);
+    this._ws.send(JSON.stringify(object));
+  },
+
+  _onMessage(event) {
+    let object = JSON.parse(event.data);
+    this.emit("onPacket", object);
+    if (this.hooks) {
+      this.hooks.onPacket(object);
+    }
+  },
+
+  close(reason) {
+    this.emit("onClosed", reason);
+
+    this.active = false;
+    this._ws.close();
+    if (this.hooks) {
+      this.hooks.onClosed(reason);
+      this.hooks = null;
+    }
+  },
+};
+
+module.exports = WebSocketDebuggerTransport;
```

Comment:
Several little things. Hopefully all can be eventually resolved.

The DOM WebSocket client (i.e., `new WebSocket("ws://localhost:6080")` can't be used, because the debugger client does some TLS magic with checking client and server certificates. That's implemented using the nsISocketTransport APIs. If we can do the same thing on the DOM WebSocket, we can migrate to it.

The event handling API is different. DOM WebSocket uses `socket.onmessage = ...`, the node.js EventEmitter (used in the `ws` module) is different from devtools EventEmitter.

Then there are differences in handling binary data. Blobs, ArrayBuffers, Arrays of ints, typed arrays, node's Buffers...